### ISSUE TITLE
Orientation information in the renderer

### ISSUE THREAD

**Issue by calintat** [homotopy-io/homotopy-rs#550]
Once the orientation labelling (#243) is done, we need to display this information in the renderered diagram.

Here are some ideas:
- different colours (how will this interact with the different shades for points, wires, regions)
- asymmetric boxes for the points (related to #519)
- arrows on the wires (#553)
- dashed/dotted wires?
- flip normals in GL

**Comment by jamievicary** [comment]
Hi -- can this please be prioritised, now we are using the alpha tool a lot it would be really great to get this rendering information in properly.

**Comment by jamievicary** [comment]
There is a problem with this idea unfortunately. Basically we cannot change the saturation, since a HSL value with positive saturation, and non-extremal lightness, will always have some "colour" to it, i.e. it will not be on the black-white spectrum. This means that, if we varied S/L as we were planning, a white or black generator would sometimes be shown with a nontrivial hue.

I think the best alternative is to just use 9 different lightness values. For example:

    L=15%, vertex, orientation -1
    L=23%, vertex, orientation 0
    L=31%, vertex, orientation +1
    L=39%, wire, orientation -1
    L=47%, wire, orientation 0
    L=55%, wire, orientation +1
    L=63%, surface, orientation -1
    L=71%, surface, orientation 0
    L=79%, surface, orientation +1

We may want to cycle these choices if the initially-chosen colour was particularly light or dark.

**Comment by calintat** [comment]
Is it really worth differentiating surfaces? For points and wires it makes sense because you can contract those, but for surfaces it seems superfluous. That would at least reduce the number of lightness values.

**Comment by jamievicary** [comment]
Yes I think it's important because you can have homotopies-on-homotopies-on-homotopies of the same generator. So you can have an entire diagram of points, wires and surfaces all with the same generator label (and potentially all with the same orientation).

**Comment by thud** [comment]
Alright I've set the lightness according to this in #717.
> L=15%, vertex, orientation -1
> L=23%, vertex, orientation 0
> L=31%, vertex, orientation +1
> L=39%, wire, orientation -1
> L=47%, wire, orientation 0
> L=55%, wire, orientation +1
> L=63%, surface, orientation -1
> L=71%, surface, orientation 0
> L=79%, surface, orientation +1

I think this is going to require a little more tweaking given that some of the
colours seem far too dark etc.
I'm going to play around with the colours and see if I can get something which
is clear but sufficiently different between codimensions/orientations.

**Comment by thud** [comment]
This is what's implemented now in #717. Note that instead of setting the lightness of a colour, I lighten or darken it by a factor.

| Co.     | Or.    | lightening fac.|
|---------|--------|----------------|
| point   | inv.   | 0.15	        |
| wire    | inv.   | 0.20	        |
| surface | inv.   | 0.25	        |
| point   | zero   | -0.05	        |
| wire    | zero   | -0.1	        |
| surface | zero   | -0.15	        |
| point   | fwd.   | 0.00	        |
| wire    | fwd.   | 0.05	        |
| surface | fwd.   | 0.10	        |

**Comment by jamievicary** [comment]
Thanks for this Jasper. I think this is much too subtle unfortunately, can you jump into Zoom for a quick discussion.

**Comment by jamievicary** [comment]
Jasper thanks for your pull request with the configurable lightnesses. I played around and a good spread of lightnesses (I used -0.36 -0.24 -0.16 -0.05 0.02 0.1 0.17 0.27 0.39) works very well regarding legibility. But we need the property we discussed before: the generator needs to be drawn in its defined colour when appearing in its "natural state", so take that as a starting point and cycle the lightnesses around the interval by adding (1/9) to achieve the 9 states. Can you do this as a priority on Monday morning please and push to alpha. Any queries please ask.

**Comment by thud** [comment]
Right now (on #717), colours are computed as follows:

|             |         | lightening factor (fixed with wrapping) |
| ----------- | ------- | :-------------------------------------: |
| positive    | point   | 0.00                                    |
| positive    | wire    | 0.08                                    |
| positive    | surface | 0.16                                    |
| zero        | point   | 0.24                                    |
| zero        | wire    | 0.32                                    |
| zero        | surface | 0.40                                    |
| negative    | point   | 0.48                                    |
| negative    | wire    | 0.56                                    |
| negative    | surface | 0.64                                    |

I know this doesn't follow the 1/9 but I personally think it achieves
good clarity.

Right now, I darken light colours instead of lightening them which avoids
wrapping in most cases (when colours wrap, they look jarringly different). I
guess the problem is that we need to make sure that colours are sufficiently
differentiable while also being recognisably the same generator.

We currently ignore the future requirement of making generators look the same
colour when viewing slices of a larger diagram.

**Comment by jamievicary** [comment]
Thanks for that update Jasper. The requirement of making generators appear their defined colour in their standard diagram is essential, of course it's ok to do it in a future PR, but I don't see how it's compatible with this plan.

Another question, what is your proposal for handling intermediate lightness colours with lightness 0.5.

**Comment by thud** [comment]
Thanks. Ok, clearly I misunderstood what was meant by natural diagram (are generators not always positive points in their natural diagram).

At the moment, colours with medium lightness are roughly treated as dark colours. There is a threshold between light and dark which is rather arbitrary (used by UI code to ensure legibility). Potentially, we could avoid wrapping any colours altogether if we make these medium lightness colours choose lightnesses either side of centre (eg. zero -> darken, positive -> darken/lighten, negative -> lighten).

**Comment by jamievicary** [comment]
So let's work out once and for all this formula for when a generator is appearing in its "natural state". An N cell appearing as a vertex in an N-diagram is in its natural state (regardless of whether we are looking at a 2d, 3d or 4d projection), as is an N-cell appearing as a wire in an (N+1)-diagram, or as a surface in an (N+2)-diagram.

So in general suppose we have an N-dimensional generator appearing in a D-dimensional diagram, where it occurs as an K-dimensional entity (K=0 for point, K=1 for wire, K=2 for surface.) Then define its codimension C = D - N - K.

Then we must colour the entity in one of 9 possible ways, depending on the 9 values of the pair (C mod 3, R) where R is the orientation number, either -1, 0 or +1.

We need the property that the pair (0, +1) is given exactly the natural colour of the generator, as that is the "standard" way the generator will be viewed.

**Comment by calintat** [comment]
I remember when I was working on hiding homotopies that it was possible for N to be larger than D, so I had to clamp D - N to make sure it doesn't underflow. I admit I now can't understand how that could happen, but that's something to keep in mind.

**Comment by doctorn** [comment]
It's easy to conceive of a 2-diagram of 4-cells (for example) so N is often greater than D

**Comment by jamievicary** [comment]
Yes the situation this can happen is if you navigate to a singular slice.

**Comment by jamievicary** [comment]
So my new suggested formula is C = S + D - N - K, where S is the total "singularness" of the slice you are viewing.

**Comment by calintat** [comment]
Yes I forgot this works on the "visible diagram" rather than the whole diagram.

**Comment by calintat** [comment]
> So my new suggested formula is C = S + D - N - K, where S is the total "singularness" of the slice you are viewing.

Hmm I think it's enough to clamp D - N - K, but I'm also not entirely sure.

**Comment by jamievicary** [comment]
OK so for now let's define C = max(0, D - N - K) and we can iterate in future if necessary.

**Comment by thud** [comment]
The recently discussed algorithm for determining colours is now implemented and merged into alpha. I'll implement this change for `gl` and `tikz` etc soon.

**Comment by jamievicary** [comment]
Thanks that is very good.

**Comment by doctorn** [comment]
Getting the colours right in 3D and 4D will require some changes to the buffering system, as, at the moment, objects of the same codimension are group by generator into draw calls and each draw call gets a single choice of colour. We'll have to key the buckets based on a (generator + orientation) composite key.

**Comment by jamievicary** [comment]
Thank you that is interesting.

**Comment by calintat** [comment]
I think we can close this, we now have colours in both 2D and 3D.

I would still like to have arrows + asymmetric boxes as an alternative to colours, but that can be tracked elsewhere.
